### PR TITLE
Super Model

### LINKED ISSUE #139: TreeModel needs refactor
The tree model should refactored to inherit from the other models (particularly RepeadedMessageModel)

### PR DESCRIPTION
**Pull request by RobertBColton**
This is a complete and total rewrite to do what #20 had proposed and resolve #139 and its ilk forever. Rather than having a bunch of repeated duplicate models with broken logic scattered all over the place, we reintroduce ProtoModel as a giant recursive super model from which the entire project can be introspected and edited. The proto model becomes one [single source of truth](https://en.wikipedia.org/wiki/Single_source_of_truth) to be reckoned with.

* Drag and drop becomes "free" for all of the editors, as the super model handles it generically for any fields.
* Only a single implementation of the main model mutation logic.
* Updates to the project are reflected across the entire editor.
* The entire project can be introspected and edited from a single diagnostic view.
* Multi-project workspaces become a lot easier to do later on.
* This makes other features like popout editor windows easier too.

Diagnostic Inspector
Inspect Sound
Diagnostic Help Menu

### CODE REVIEW

**Comment by JoshDreamland** [comment]
Let's revisit this. I need to sit down and piece together what's going on here. I'm told it's unfinished, but I like the 1KLoC reduction in size, for sure. Resurrecting this may be more efficient than what I was about to do (might end up going down the same rabbit holes).

**Comment by RobertBColton** [comment]
The main issue here Josh is with change propagation. While the current/other way has issues propagating changes up and out to the rest of the editor, this way has the problem that it propagates changes too often. Something as simple as changing the name of a folder, would recursively signal everything below it as changed. That was why I decided that the current approach of model composition was best, because it's easier to deal with bubbling up than bubbling down.

In other words, this pull request is essentially proposing Vuex for Qt. They talk about this in their intro video.
https://vuex.vuejs.org/
